controllers/systemconfiguration_controller.go

```diff
@@ -0,0 +1,164 @@
+	"github.com/DataWorkflowServices/dws/internal/controller/metrics"
+	"github.com/DataWorkflowServices/dws/utils/updater"
+	"github.com/go-logr/logr"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+const (
+	// finalizerDWSSystemConfiguration defines the finalizer name that this controller
+	// uses on the SystemConfiguration resource. This prevents the SystemConfiguration resource
+	// from being fully deleted until this controller removes the finalizer.
+	finalizerDWSSystemConfiguration = "dataworkflowservices.github.io/systemconfiguration"
+)
+
+// SystemConfigurationReconciler reconciles a SystemConfiguration object
+type SystemConfigurationReconciler struct {
+	client.Client
+	Log          logr.Logger
+	Scheme       *runtime.Scheme
+	ChildObjects []dwsv1alpha2.ObjectList
+}
+
+// +kubebuilder:rbac:groups=dataworkflowservices.github.io,resources=systemconfigurations,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=dataworkflowservices.github.io,resources=systemconfigurations/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=dataworkflowservices.github.io,resources=systemconfigurations/finalizers,verbs=update
+// +kubebuilder:rbac:groups=dataworkflowservices.github.io,resources=storages,verbs=get;list;watch;create;update;patch;delete;deletecollection
+
+func (r *SystemConfigurationReconciler) Reconcile(ctx context.Context, req ctrl.Request) (res ctrl.Result, err error) {
+	log := r.Log.WithValues("SystemConfiguration", req.NamespacedName)
+
+	metrics.DwsReconcilesTotal.Inc()
+
+	systemConfiguration := &dwsv1alpha2.SystemConfiguration{}
+	if err := r.Get(ctx, req.NamespacedName, systemConfiguration); err != nil {
+		// ignore not-found errors, since they can't be fixed by an immediate
+		// requeue (we'll need to wait for a new notification), and we can get them
+		// on deleted requests.
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	// Create a status updater that handles the call to r.Status().Update() if any of the fields
+	// in systemConfiguration.Status{} change
+	statusUpdater := updater.NewStatusUpdater[*dwsv1alpha2.SystemConfigurationStatus](systemConfiguration)
+	defer func() { err = statusUpdater.CloseWithStatusUpdate(ctx, r.Client.Status(), err) }()
+	defer func() { systemConfiguration.Status.SetResourceErrorAndLog(err, log) }()
+
+	// Handle cleanup if the resource is being deleted
+	if !systemConfiguration.GetDeletionTimestamp().IsZero() {
+		if !controllerutil.ContainsFinalizer(systemConfiguration, finalizerDWSSystemConfiguration) {
+			return ctrl.Result{}, nil
+		}
+
+		deleteStatus, err := dwsv1alpha2.DeleteChildren(ctx, r.Client, r.ChildObjects, systemConfiguration)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+
+		if !deleteStatus.Complete() {
+			return ctrl.Result{}, nil
```

Comment:
Should this ask for a requeue?

Reply:
It will get an event when the Storage resource disappears. The other controllers are set up this way too.